### ISSUE TITLE
Add a way to replace a track's master audio

### ISSUE THREAD

**Issue by Feld0** [Poniverse/Pony.fm#9]
This should be available through the standard track editor and the eventual admin tools (#2). It would both allow a track to be updated if the artist wishes, and facilitate the replacement of lossy masters with lossless ones in the future.

**Comment by kz** [comment]
@Feld0 For the backend command, what would the method of input be? A remote URL to download the new master file from?

**Comment by Feld0** [comment]
@kz I'd go with an `UploadedFile` object - the same kind as what normally gets returned by `Input::file()`.

**Comment by kz** [comment]
@Feld0 Could you elaborate on that? My question is how to get the file on the server in the first place for the backend/console command. Unless by backend, you mean admin panel?

**Comment by Feld0** [comment]
@kz I thought you were referring to the API for the "ReplaceTrackMasterAudio" command.

When all is said and done, a new audio file should be uploaded by the user. The existing uploader should be adapted for that and AFAIK, @logic-dev was going to help with the front-end part of this.

If you're only working on the backend at this point, I suggest looking for an audio file at a hardcoded path (the `storage/app` directory would be appropriate, since you can easily drop a file into it from your host), which will be replaced by a user-uploaded file once the front-end part is in place.

**Comment by logic-dev** [comment]
Yep, once that backend API is in place I'll add it to the front end

**Comment by kz** [comment]
Would it be a bad move to create a command which piggybacks off UploadTrackCommand (which creates a new track, generates the relevant track files and tags), soft-delete the newly-uploaded track and then switch over the track-files to the old track? This would be accompanied by a table which logs all revisions to an original track.

It appears that GenerateTrackFilesCommand needs a $track and deletes the $track passed to it if generation fails. As a result, I can't pass in the original track to the command, so it would be more straightforward to generate a new one by using UploadTrackCommand.

**Comment by Feld0** [comment]
@kz Track ID's should be preserved across revisions. The `tracks` table should have one row for every logical track, regardless of how many revisions its audio file has.

Idea: what if a new `version` column was added to the `track_files` table, which starts at 1 for every track, and a new set of `track_files` records was created for every new revision (with an incremented `version` value)? The "current" downloads for a track would be the set of track files with the largest `version` number. Older versions of a track's files would, with the exception of the master file, be automatically expired.

`GenerateTrackFilesCommand` could be adapted to generate a new set of track files whenever it's called.

**Comment by kz** [comment]
@Feld0 I'm currently working on the command which handles an uploaded file and passes it to GenerateTrackFilesCommand. One thing to implement is the deletion of non-master files. At what point of the process would be best to handle this? 

One option would be in [EncodeTrackFile.php](https://github.com/Poniverse/Pony.fm/blob/3b1b03e12baf4ffa4c9affd5685d5e0bec915780/app/Jobs/EncodeTrackFile.php#L161-L161) after it handles `autoPublish`. Like with `autoPublish`, a new parameter (`deletePreviousVersionFiles`?) needs to be added for the constructor. However, does that seem appropriate for what `EncodeTrackFile` handles?

Another option would be to run an artisan command every n hours to delete non-master files for previous versions. An advantage to this would be that if the user does decide to revert before the n hours are up, the files would still be there.

**Comment by kz** [comment]
Another thing I've been thinking about is how to handle the act of reverting versions. My current thought is to add a `current_version` column to `tracks`, let artists upload a number of versions (perhaps impose a limit?) and allow them to switch between versions (a `use` button) and of course, delete them. As a result, the artist can use v2 for a track even if they have v3+ uploaded.

Alternatively, we could have a more linear versioning system, where reverting from v3 to v1 deletes the track files for v2 and v3.

**Comment by Feld0** [comment]
@kz Deleting old non-master files fits most neatly into `ClearTrackFileCacheCommand` IMO (it may actually have been called something else; that's off the top of my head). There's no particular rush to expire old non-master files, and the Artisan command for clearing the track file cache is already in place. This fits the spirit of purging unneeded track files that we are able to regenerate.

Reverting to a version that's queued for deletion could actually create a nasty race condition that you'll have to look out for.

I like your `current_version` idea. Note that revisions should not be possible to delete.

**Comment by kz** [comment]
I've been doing some thinking and have a proposal of how to approach the task and would appreciate any comments.

Database: 

- add `version_upload_status` to `tracks` table
- add `current_version` to `tracks` table
- add `version` to `track_files` with default `1`

Storage:

- create a new folder to store the master of all versions which are not currently being used, from now on referred to as the 'master store'

UploadVersionCommand:

- `$track->version_upload_status` set to `Track::STATUS_VERSION_UPLOADING`
- Run EncodeVersionCommand (async)

EncodeVersionCommand:

- similar to EncodeTrackFileCommand, however is adapted to track versions
- constructor takes `$track` (Track instance), `$masterPath` (either uploaded file in temp or a file from the master store) and `$newVersion` (the version to change `$track->version` to if encoding is successful)
- the master in the $masterPath is taken and all files are encoded to the temp folder
- on success:
  - the original master file is moved to the master store and the remaining original files are deleted
  - the encoded files are moved from the temp folder to the track files folder
  - new TrackFile rows are created with `$newVersion` if they do not already exist and `$track->current_version` is set to `$newVersion`
  - file sizes are repopulated
  - `$track->version_upload_status` set to `Track::STATUS_VERSION_SUCCESS`
- on failure (during the encoding stage):
  - all temporary files deleted
  - `$track->version_upload_status` set to `Track::STATUS_VERSION_FAILED`

Method to choose version:

- run EncodeVersionCommand with the path to the version in the master store

Method to view status of the encoding

- returns `$track->version_upload_status`

For the methods, should they be under `TracksController` or a new controller (`TrackVersionsController`)?

**Comment by kz** [comment]
@Feld0 With the above plan, it wouldn't be possible to keep the original files where they are since the newly encoded files for a new version would obviously have to replace the original files. Would you prefer to immediately delete the original files, or move them elsewhere so they can be moved back without re-encoding if an artist decides to revert to the previous version before the artisan command is run?

**Comment by Feld0** [comment]
@kz

**Database:** those changes look good.

**Storage:** I'd rather we not make a habit of moving files around whenever the version number changes - or moving them around at all after they're generated. Consider changing the naming scheme for track files instead to, say, `{track_id}-{version}.{extension}` and relying on the existing caching + master file preservation mechanism to keep master files intact. This naming scheme change can be safely performed with a database migration.

**UploadVersionCommand:** this sounds similar enough to `UploadTrackCommand` that they can probably be merged. `UploadTrackCommand` could take a track object as an optional parameter which, if given, will attach the new version to the track instead of making a new track.

**EncodeVersionCommand:** the `masterPath` argument is not needed; we should be able to generate the path to all track files deterministically (and their records will already exist in the `track_files` table before this is called). I'm not convinced this command needs to be a separate one from `EncodeTrackFileCommand`, as the only functionality in it that differs from `EncodeTrackFileCommand` is that it changes the version number when all of a particular version's encodes complete successfully.

Track versions should be considered, as much as possible, an implementation detail of tracks; I'd include any API endpoints specific to track versions under the `TracksController`.

**Comment by kz** [comment]
Thanks for the review. Will start now.

Task list:
- [ ] Add migrations
- [ ] Update naming scheme by modifying models
- [ ] Create command to rename naming scheme for all track files
- [ ] Modify `EncodeTrackFileCommand` to encode new versions of files
- [ ] Modify `UploadTrackCommand` to handle track uploading
- [ ] Modify `TracksController` to add methods for an API

**Comment by kz** [comment]
The main commit is done now. Just doing final checking to make sure albums, playlists and the console command work as expected.

**Comment by logic-dev** [comment]
Merged the backend into master. Will deploy on prod soon. Once deployed and verified on prod I'll start working on the front end

**Comment by logic-dev** [comment]
Basic front end for admins has been deployed. User facing will come once polished
